comms/dht/src/crypt.rs

```diff
@@ -20,29 +20,55 @@
 // WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 // USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+use chacha20::{ChaCha20, Key, Nonce};
+use chacha20::cipher::{NewCipher, StreamCipher};
+use rand::{rngs::OsRng, RngCore};
 use tari_comms::types::CommsPublicKey;
 use tari_crypto::{
     keys::{DiffieHellmanSharedSecret, PublicKey},
-    tari_utilities::{
-        ciphers::{
-            chacha20::ChaCha20,
-            cipher::{Cipher, CipherError},
-        },
-        ByteArray,
-    },
+    tari_utilities::ByteArray,
 };
+use crate::outbound::DhtOutboundError;
+use std::mem::size_of;
 
 pub fn generate_ecdh_secret<PK>(secret_key: &PK::K, public_key: &PK) -> PK
 where PK: PublicKey + DiffieHellmanSharedSecret<PK = PK> {
     PK::shared_secret(secret_key, public_key)
 }
 
-pub fn decrypt(cipher_key: &CommsPublicKey, cipher_text: &[u8]) -> Result<Vec<u8>, CipherError> {
-    ChaCha20::open_with_integral_nonce(&cipher_text.to_vec(), cipher_key.as_bytes())
+pub fn decrypt(cipher_key: &CommsPublicKey, cipher_text: &[u8]) -> Result<Vec<u8>, DhtOutboundError> {
```

Comment:
Is decypting with a pubkey correct? How is the en- / decrpytion used in the comms? If it's all with pubkeys, then the encryption is likely meaningless.

Reply:
Its a Diffie-Hellman key

Reply:
The type here is confusing. Perhaps just a different type?

Reply:
DH keys are typically secret keys. Yeah, it's just confusing.

Reply:
Would need to update the way its implemented in the `tari_crypto` crate. 
https://github.com/tari-project/tari-crypto/blob/main/src/keys.rs#L78

Reply:
Yes, it's incorrect in tari_crypto. @hansieodendaal picked that up a few weeks ago and was going to push a PR changing it.